internal/provider/provider.go

```diff
@@ -16,5 +24,99 @@ func New() *schema.Provider {
 			"tls_public_key":  dataSourcePublicKey(),
 			"tls_certificate": dataSourceCertificate(),
 		},
+		Schema: map[string]*schema.Schema{
+			"proxy": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"url": {
+							Type:             schema.TypeString,
+							Optional:         true,
+							ValidateDiagFunc: validation.ToDiagFunc(validation.IsURLWithScheme(SupportedProxySchemesStr())),
+							ConflictsWith:    []string{"proxy.0.from_env"},
+							Description: "URL used to connect to the Proxy. " +
+								fmt.Sprintf("Accepted schemes are: `%s`. ", strings.Join(SupportedProxySchemesStr(), "`, `")),
+						},
+						"username": {
+							Type:         schema.TypeString,
+							Optional:     true,
+							RequiredWith: []string{"proxy.0.url"},
+							Description:  "Username (or Token) used for Basic authentication against the Proxy.",
+						},
+						"password": {
+							Type:         schema.TypeString,
+							Optional:     true,
+							Sensitive:    true,
+							RequiredWith: []string{"proxy.0.username"},
+							Description:  "Password used for Basic authentication against the Proxy.",
+						},
+						"from_env": {
+							Type:          schema.TypeBool,
+							Optional:      true,
+							Default:       false,
+							ConflictsWith: []string{"proxy.0.url", "proxy.0.username", "proxy.0.password"},
+							Description: "When `true` the provider will discover the proxy configuration from environment variables. " +
+								"This is based upon [`http.ProxyFromEnvironment`](https://pkg.go.dev/net/http#ProxyFromEnvironment) " +
+								"and it supports the same environment variables (default: `true`).",
+						},
```

Comment:
For the first item, the HTTP server might need a way to reject connections except those from the proxy.

Reply:
1. will add tests to use `t.Setenv` 👍 
2. Will fix the description
3. I'm not sure, my intention wasn't to make the `from_env` default: are you suggesting it should be made the default in a major release?
4. I can look into ways to orchestrate a "reject if not via proxy": I can try to play with headers injection more, but I have noticed that, for what our provider does, we never get to exchange http body - so requests never actually get to exchange headers. Will give it another pass.

Reply:
> my intention wasn't to make the from_env default: are you suggesting it should be made the default in a major release?

I think we should certainly consider this for the next major version. 😄  Many pieces of software handling HTTP calls will conventionally support the `NO_PROXY`/`HTTP_PROXY`/`HTTPS_PROXY` environment variables without additional configuration. github.com/hashicorp/go-cleanhttp default clients for example, implement net/http.ProxyFromEnvironment, which I believe is used by the majority of HTTP handling within Terraform CLI itself.

Reply:
So, the last point you make is great, but it kinda goes against the other thing we decided: to not use an HTTP client unless a Proxy is configured.

Using `ProxyFromEnvironment` by default would also change the current behaviour.

So, I think now I follow better what you head in mind earlier when you suggested:
> Should we note that this will be enabled by default in the next major version and add a code comment so that gets done?

In the light of this, I think I'll do the following:
* by default, if no `proxy` config is provided, we keep using `tls.Client`
* if a proxy is configured, then we switch the logic to use `http.Client`
* `from_env` is defaulted to `false`

Then, I'll create a follow up issue for breaking changes that we want to launch in 4.x (similar to [this one](https://github.com/hashicorp/terraform-provider-tls/issues/174)). In this issue I'll include:
* `from_env` becomes defaulted to `true`
* the URL scheme defines the client we will use: this means `https://` will always use the `http.Client`, while `tls://` will always use `tls.Client`

Does this strategy sound good? Am I missing something?